Added SSH resource

mjtravers opened the pull request:
I had to comment out one of the tests in interface/tests/test_retrace_packages.py to get my test to run. I put a "TODO" comment in the file to document the change.

Also,  passwords are currently lost between niceman calls to the same resource. I didn't think we should save it in the inventory file. I left it as a "TODO" as well to be worked out later.

yarikoptic commented:
Can't check on the laptop but smells like you rebased or smth like that, not just merged master... Is that the case?

mjtravers commented:
Yes, I merged master into my own master and then rebased my enh_ssh on the
updated master.

Was there a better way to do the merge?

On Fri, Sep 1, 2017 at 6:09 PM, Yaroslav Halchenko <notifications@github.com
> wrote:

> Can't check on the laptop but smells like you rebased or smth like that,
> not just merged master... Is that the case?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ReproNim/niceman/pull/103#issuecomment-326694163>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AC1YiTeF43OZVMJ2sYMNe9xgu--D6rBJks5seICcgaJpZM4O8Ugw>
> .
>



-- 
*Matt Travers,* Software Developer
*TCG, Inc. - Positively Distinct* - CMMI-DEV Level 3 - CMMI-SVC Level 2 -
ISO 9001:2015
703-946-5935 | matt.travers@tcg.com | www.tcg.com
<https://www.facebook.com/TCG-32241785903>  <https://twitter.com/TCGnews>

yarikoptic commented:
If you already merged, why bother rebasing? Do one, or another.

yarikoptic commented:
fetched your clone locally... I do not see anything being rebased after the merge, so should be all kosher -- not sure why github listed those commits which are now part of master... although part of the confusion here is seems merging two trees where one was rebased before I guess:
```shell
$> git lg master...gh-mjtravers/enh_ssh 
*   dea9b48 - (HEAD, gh-mjtravers/enh_ssh) Merge branch 'enh_ssh' of https://github.com/mjtravers/niceman into enh_ssh (16 hours ago) [Matt Travers]
|\  
| * 7e37fa0 - ENH: Moved SSH connection session out of execute command methods. (2 weeks ago) [Matt Travers]
| * 92cebc1 - BF: Correcting failing tests. (2 weeks ago) [Matt Travers]
| * 3ed2b97 - Initial commit for addition of SSH resource. (2 weeks ago) [Matt Travers]
* 3bf74ff - ENH: Moved SSH connection session out of execute command methods. (16 hours ago) [Matt Travers]
* e5cace5 - BF: Correcting failing tests. (16 hours ago) [Matt Travers]
* ddd280e - Initial commit for addition of SSH resource. (16 hours ago) [Matt Travers]
```
I guess -- just fix up for the travis now and we would not dive too deep into all this ;)  but in general, either rebase, or merge... rarely need both together I think ;)

mjtravers commented:
Hmmmmm, I rebased my enh_ssh branch after I merged the master branches so I
could rectify the conflicts from the sessions branch PR merge. Just merging
the masters left the enh_ssh branch pointing to the original old commit
without the new session code.

Multiple branch/repo synchronizing mystifies me. ☺

On Sat, Sep 2, 2017 at 10:09 AM, Yaroslav Halchenko <
notifications@github.com> wrote:

> fetched your clone locally... I do not see anything being rebased after
> the merge, so should be all kosher -- not sure why github listed those
> commits which are now part of master... although part of the confusion here
> is seems merging two trees where one was rebased before I guess:
>
> $> git lg master...gh-mjtravers/enh_ssh *   dea9b48 - (HEAD, gh-mjtravers/enh_ssh) Merge branch 'enh_ssh' of https://github.com/mjtravers/niceman into enh_ssh (16 hours ago) [Matt Travers]|\  | * 7e37fa0 - ENH: Moved SSH connection session out of execute command methods. (2 weeks ago) [Matt Travers]| * 92cebc1 - BF: Correcting failing tests. (2 weeks ago) [Matt Travers]| * 3ed2b97 - Initial commit for addition of SSH resource. (2 weeks ago) [Matt Travers]* 3bf74ff - ENH: Moved SSH connection session out of execute command methods. (16 hours ago) [Matt Travers]* e5cace5 - BF: Correcting failing tests. (16 hours ago) [Matt Travers]* ddd280e - Initial commit for addition of SSH resource. (16 hours ago) [Matt Travers]
>
> I guess -- just fix up for the travis now and we would not dive too deep
> into all this ;) but in general, either rebase, or merge... rarely need
> both together I think ;)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ReproNim/niceman/pull/103#issuecomment-326746169>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AC1Yif3VH2NDx8tKxakEmm2zfUZ4QLOfks5seWGMgaJpZM4O8Ugw>
> .
>



-- 
*Matt Travers,* Software Developer
*TCG, Inc. - Positively Distinct* - CMMI-DEV Level 3 - CMMI-SVC Level 2 -
ISO 9001:2015
703-946-5935 | matt.travers@tcg.com | www.tcg.com
<https://www.facebook.com/TCG-32241785903>  <https://twitter.com/TCGnews>

yarikoptic commented:
Don't remember which of the two sure in top, but I think that after you rebased on top of master, you were all set, and that is why merge of master didn't change anything since master was already past of your history. But then you also merged your older version (before rebasing) which should have resulted in no changes really but just brought two trees together. Hope this helps. Using gitk usually helps me to visually inspect and understand what is going on ;-)

yarikoptic commented on niceman/interface/tests/test_retrace_packages.py:
```diff
@@ -33,14 +33,15 @@ def test_identify_packages():
-    assert distributions[0].packages[0].files == [__file__]
-
-    assert files == ['/nonexisting-for-sure']
+# TODO: This test case is failing in my dev environment due to local paths.
```
how are your paths so special?
ideally we should just get it fixed, instead of disabling tests

yarikoptic commented on niceman/resource/base.py:
```diff
@@ -11,6 +11,10 @@
+try:
+    from ConfigParser import NoSectionError
+except:
+    from configparser import NoSectionError
```
in such cases, just check if six does not provide needed "moves", in this case it does:
```from six.moves.configparser import NoSectionError```

mjtravers commented:
I put back in that test and added the six package in for the NoSectionError excpetion.

six.move is good to know

yarikoptic commented on niceman/interface/ls.py:
```diff
@@ -12,6 +12,10 @@
+try:
+    from ConfigParser import NoSectionError
+except:
+    from configparser import NoSectionError
```
The same six

yarikoptic commented on niceman/resource/ssh.py:
```diff
@@ -0,0 +1,132 @@
+    def stop(self):
+        return
+
+    def execute_command(self, command, env=None):
```
I think this now should be moved into `SSHSession(POSIXSession)` class under `_execute_command`, and `get_session` should be implemented here?

yarikoptic commented on niceman/resource/ssh.py:
```diff
@@ -0,0 +1,132 @@
+            lgr.info("Running command '%s'", command['command'])
+            self.execute_command(command['command'], command['env'])
+
+    def login(self):
```
this should move away and become a `PTYSSHSession.open()`? see how done for Docker

mjtravers commented on niceman/support/sshconnector2.py:
```diff
@@ -21,6 +21,16 @@ class SSHConnector2(object):
     Manage and use an SSH connection.
     """
 
+    def connect(self):
```
Since we are using the SSH client from starcluster for the PTY connections, we can also use it for executing shell commands on the remote server and drop the use of 2 different SSH libraries by eliminating the sshconnector2.py package.

yarikoptic commented on .travis.yml:
```diff
@@ -50,6 +53,8 @@ before_install:
   - git config --global user.email "test@travis.land"
   - git config --global user.name "Travis Almighty"
+  # Setup SSH docker service
+  - docker run -d -p 49000:22 rastasheep/ubuntu-sshd:14.04
```
I think travis does not support docker ... we would need to use circle-ci if we decide to test against live docker instances

mjtravers commented:
This PR is ready for another review. The session code for the AWS resource is now using the SSH session code.

yarikoptic commented on niceman/resource/ssh.py:
```diff
@@ -0,0 +1,144 @@
+        assert self._ssh, "We should create or connect to remote server first"
+
+        if pty and shared is not None and not shared:
+            lgr.warning("Cannot do non-shared pty session for ssh server yet")
```
are you sure that we are getting only shared ones?  I think it is the opposite (I could be wrong0

yarikoptic commented on niceman/resource/ssh.py:
```diff
@@ -0,0 +1,144 @@
+
+        # If a command fails, a CommandError exception will be thrown.
+        for i, line in enumerate(self.ssh.execute(" ".join(command))):
+            lgr.debug("exec#%i: %s", i, line.rstrip())
```
also we will need other methods Session should get but POSIXSession (or actually atm none apparently) define -- `copy_to` and `copy_from` so we could transfer needed files back and forth

yarikoptic commented on niceman/resource/tests/test_ssh.py:
```diff
@@ -0,0 +1,53 @@
+        resource = ResourceManager.factory(config)
+        updated_config = resource.create()
+        config.update(updated_config)
+        assert re.match('\w{8}-\w{4}-\w{4}-\w{4}-\w{12}',
```
you might want to add a trailing anchor `$` since apparently (as I have discovered after years of blessed ignorance) `match` just guarantees matching from the beginning... e.g.
```python
$> python -c 'import re; print(re.match("\w{4}", "toolong"))'         
<_sre.SRE_Match object at 0x7f50b7980718>

$> python -c 'import re; print(re.match("\w{4}$", "toolong"))' 
None
```

mjtravers commented:
Okay, ready for another code review. I made updates reflecting the last round of comments.

yarikoptic commented on niceman/resource/ssh.py:
```diff
@@ -0,0 +1,189 @@
+        -------
+        dict : config and state parameters to capture in the inventory file
+        """
+        self.id = str(uuid.uuid4())
```
should may be done only `if not self.id`?  e.g. if we created one originally with some id, shouldn't get overrident

yarikoptic commented on niceman/resource/ssh.py:
```diff
@@ -0,0 +1,189 @@
+        return
+
+    def stop(self):
+        return
```
shouldn't we explicitly `self._ssh=None` or smth?

mjtravers replied:
I put the self._ssh=None in the "delete" method. The stop may be followed by a start sometime.

yarikoptic commented on niceman/resource/ssh.py:
```diff
@@ -0,0 +1,189 @@
+        """
+        Log into remote environment and get the command line
+        """
+        assert self._ssh, "We should create or connect to remote server first"
```
so why not to try to do it automagically?  `.connect` doesn't require any option

yarikoptic commented on niceman/resource/ssh.py:
```diff
@@ -0,0 +1,189 @@
+            ssh=self._ssh
+        )
+
+# Alias SSH class so that it can be discovered by the ResourceManager.
```
any why otherwise it (`SSH`) is not discovered?

mjtravers replied:
Right, it will not find the module because the discovery process is looking for CamelCase class names based on the hyphen-delineated resource types declared.

e.g. docker-container looks for class DockerContainer

likewise, a resource of type "ssh" is looking for a class "Ssh"

yarikoptic commented on niceman/resource/ssh.py:
```diff
@@ -0,0 +1,189 @@
+                owner=None, group=None, recursive=False):
+        """Take file on the local file system and copy over into the session
+        """
+        self.ssh.put([src_path], remotepath=dest_path)
```
anything about `preserve_perms`?

yarikoptic replied:
what does dest_dir='.' means ?  current directory (e.g. /home/yoh/proj/repronim/niceman/...) but on remote host?

yarikoptic commented on niceman/resource/ssh.py:
```diff
@@ -0,0 +1,189 @@
+                  owner=None, group=None, recursive=False):
+        """Retrieve a file from the remote system
+        """
+        self.ssh.get(src_path, localpath=dest_path)
```
in both above -- will it be recursive or not?

yarikoptic commented on niceman/resource/ssh.py:
```diff
@@ -0,0 +1,189 @@
+        """Create a directory (or with parent directories if `parents`
+        is True)
+        """
+        self.ssh.mkdir(path, mode=0o755)
```
what about `parents`?

yarikoptic commented on niceman/resource/ssh.py:
```diff
@@ -0,0 +1,189 @@
+
+
+
+
```
2... or 3 empty lines should be sufficient I guess ;)

yarikoptic commented on niceman/resource/tests/test_ssh.py:
```diff
@@ -0,0 +1,76 @@
+
+        session.mkdir('test-dir')
+        assert session.isdir('test-dir') == True
+        assert session.isdir('not-a-dir') == False
```
moreover it doesn't even exist.  what about checking also `/etc/hosts` to be not a dir?

yarikoptic commented on niceman/resource/tests/test_ssh.py:
```diff
@@ -0,0 +1,76 @@
+        assert session.exists('/no/such/file') == False
+
+        session.copy_to(__file__)
+        assert session.exists('test_ssh.py') == True
```
well -- if it copies simply to the same location (warning: could destroy an original if somehow fails... good it is under git at least), then this check isn't really checking that transfer was succesfull.

yarikoptic commented on niceman/support/starcluster/sshutils.py:
```diff
@@ -410,6 +411,7 @@ def progress_bar(self):
         return self._progress_bar
 
     def _file_transfer_progress(self, filename, size, sent):
+        pass
```
why do we need this change?

yarikoptic commented:
I am afraid we are piling up conflicts, and this one seems to be more or less ready, so I will merge it. Cheers!
